### PR TITLE
[js/web] allow ShaderHelper to use internal (non-I/O) variables

### PR DESCRIPTION
**Pull request by fs-eire**
### Description
This PR includes a change that inspired from #18452 to resolve a requirement: a shader may depend on an instance of `IndicesHelper` to generate WGSL code snippet, but the IndicesHelper instance is not necessarily an input/output of the program. So the existing `declareVariables()` function does not work with this scenario.

In order to support this requirement, I added this "use" function to `interface ShaderHelper`, which takes a helper-like object as parameter. The hidden implementation `ShaderHelperImpl` class will iterate the helpers and call `impl()` for each.

@axinging @qjia7

### CODE REVIEW

**Review by qjia7 — APPROVED** [review]
LGTM, thanks for the solution!

**Comment by axinging** [comment]
With this change, program should make below change to implement offsetToIndices?
```
const batchDims = inputVariable('batchDims', inputs[0].dataType, batchShapeOrRank, 1); 
shaderHelper.use(() =>batchDims.offsetToIndices('u32(batch)'));
const inputVariables = [batchDims, A, B];
shaderHelper.registerUniform('dimBOuter', 'i32')
              .registerUniform('dimInner', 'i32')
              .declareVariables(...inputVariables, output)}
```

**Comment by qjia7** [comment]
> With this change, program should make below change to implement offsetToIndices?
> 
> ```
> const batchDims = inputVariable('batchDims', inputs[0].dataType, batchShapeOrRank, 1); 
> const inputVariables = [batchDims, A, B];
> shaderHelper.registerUniform('dimBOuter', 'i32')
>               .registerUniform('dimInner', 'i32').use(() =>batchDims.offsetToIndices('u32(batch)'))
>               .declareVariables(...inputVariables, output)}
> ```

How do you think we change it like below: extend `IndicesHelper`'s usage to `'input'|'output'|'internal'` from `'input'|'output'`. Then
```
const batchDims = internalVariable('batchDims', inputs[0].dataType, batchShapeOrRank)
const inputVariables = [A, B];
 shaderHelper.registerUniform('dimBOuter', 'i32')
               .registerUniform('dimInner', 'i32')
               .registerInternalVariables(batchDims).
               .declareVariables(...inputVariables, output)}

**Comment by fs-eire** [comment]
Will think about making corresponding changes later.

File: js/web/lib/wasm/jsep/webgpu/ops/common.ts [hunk 1]
```diff
@@ -785,7 +855,8 @@ class ShaderHelperImpl implements ShaderHelper {
    * Get additional implementation that needs to be added to the shader source.
    */
   get additionalImplementations(): string {
-    return this.uniformDeclaration() + this.indicesHelpers.map(i => i.impl()).join('\n');
+    return this.uniformDeclaration() + this.variables.map(i => i.impl()).join('\n') +
+        this.helpers.map(i => i.impl()).join('\n');
```
Comment (qjia7) [inline js/web/lib/wasm/jsep/webgpu/ops/common.ts:859]: Does it work if we remove `use` and `helpers` and directly call like below?
```
return this.uniformDeclaration() + this.variables.map(i => i.impl()).join('\n') + this.internalVariables.map(i => i.impl()).join('\n');

Comment (fs-eire) [reply js/web/lib/wasm/jsep/webgpu/ops/common.ts:859]: let me remove 'use'

**Comment by fs-eire** [comment]
Now removed `use()` and port on `registerInternalVariables()`.

**Review by qjia7 — APPROVED** [review]
Nice 👍 

Do you need to update this PR's title and description?

**Comment by fs-eire** [comment]
merged to latest main branch.
@qjia7 @axinging

**Review by satyajandhyala — APPROVED** [review]

**Review by guschmue — APPROVED** [review]

**Pull request merged**